### PR TITLE
Single-language, single-tab app translations upload

### PR DESCRIPTION
**Pull request by orangejenny**
[JIRA ticket](https://dimagi-dev.atlassian.net/browse/ICDS-295)

[QA ticket](https://dimagi-dev.atlassian.net/browse/QA-264)

[Original spec](https://docs.google.com/document/d/1ZAmUE7bHRmsNx6EYWcuE4pmu1eAm4LwDEkvb_8_Rzak/), although it has deviated a bit. The general idea is to make the download added in https://github.com/dimagi/commcare-hq/pull/23391 be able to be re-uploaded.

The UI has changed: this isn't really a multimedia-specific tool, even though ICDS is planning to only use it to update audio files, so it's now a part of the bulk app trans section instead of being its own section. If you have the flag turned on, you see the language dropdown, and if you pick a language, you get the single-tab file.

Screen Shot 2019-03-14 at 8 30 40 PM

Review by commit. Many of the commits just extract helper functions and split `app_translations.py` into multiple files. It might be helpful to read the [README](https://github.com/dimagi/commcare-hq/blob/e3ca07e8e80728e80e959a3b6be3fb8e8f7ae9b9/corehq/apps/translations/app_translations/README.md) before reviewing.

Product: feature flag is "Bulk multimedia path management." This does make also minor changes to the all-users-facing bulk app translation, like reordering columns and changing the names of the sheets. All of these changes should be backwards compatible, so uploading a pre-release file should work. I'll remove the backwards compatibility a few days after release.

### CODE REVIEW

File: corehq/apps/translations/views.py
Comment (czue) [inline corehq/apps/translations/views.py]: when is workbook empty?

File: corehq/apps/translations/app_translations.py
Comment (czue) [inline corehq/apps/translations/app_translations.py]: this is a great change. one thought is that now that it's extracted to its own function you could manage control flow with exceptions which might be a little more explicit than relying on the return value being `None`

e.g. 
```python
try:
  _check_for_sheet_error(...)
except InvalidSheetException as e:
  msgs.append(e.message)
```

File: corehq/apps/translations/app_translations.py
Comment (czue) [inline corehq/apps/translations/app_translations.py]: nit: comment should probably be under this line

File: corehq/apps/translations/app_translations.py
Comment (czue) [inline corehq/apps/translations/app_translations.py]: curious how this still works since it seems like `get_unicode_dicts(sheet)` now gets called multiple times?

File: corehq/apps/translations/app_translations.py
Comment (czue) [inline corehq/apps/translations/app_translations.py]: could make this a `namedtuple` (is very low overhead and a just bit more explicit)

File: corehq/apps/translations/app_translations.py
Comment (czue) [inline corehq/apps/translations/app_translations.py]: same comment (especially re: commit message) about `namedtuple` could apply here

File: corehq/apps/translations/app_translations/upload_module.py
Comment (czue) [inline corehq/apps/translations/app_translations/upload_module.py]: repeating myself a lot but could be `namedtuple` here too. though also realize this was just moved.

**Review by czue — COMMENTED** [review]
Looks great overall and like a really nice cleanup. I had a few small non-blocking suggestions and one question.

What is our test coverage for these different pathways and uploads? It would be great if we could get reasonable test coverage for the different formats and error cases.

File: corehq/apps/translations/app_translations/upload_module.py
Comment (kaapstorm) [inline corehq/apps/translations/app_translations/upload_module.py]: I know you didn't write this docstring, but is it intentionally not a doctest? 

If you change `>>` to `>>>`, `true` to `True` and `false` to `False`, then you can get the test suite to [check the docstring](https://github.com/dimagi/commcare-hq/blob/868436c0a4c1a4daaeefd6c89ef6310af3d542e7/corehq/sql_db/tests/test_backports.py#L50) is correct with 

```python
def test_doctests():
    results = doctest.testmod(upload_module)
    assert results.failed == 0
```

File: corehq/apps/translations/app_translations/README.md
Comment (kaapstorm) [inline corehq/apps/translations/app_translations/README.md]: `generates generates`; nit: "Excel" is a proper noun.

File: corehq/apps/translations/app_translations/README.md
Comment (kaapstorm) [inline corehq/apps/translations/app_translations/README.md]: nitto

File: corehq/apps/translations/app_translations/upload_app.py
Comment (kaapstorm) [inline corehq/apps/translations/app_translations/upload_app.py]: nit: Docstring no longer matches function signature.

**Review by kaapstorm — APPROVED** [review]
I love the refactors.

File: corehq/apps/translations/app_translations.py
Comment (mkangia) [inline corehq/apps/translations/app_translations.py]: i'd just call this `validate_sheets` or `validate_workbook`.

File: corehq/apps/translations/app_translations.py
Comment (mkangia) [inline corehq/apps/translations/app_translations.py]: why this change?

File: corehq/apps/translations/app_translations/README.md
Comment (mkangia) [inline corehq/apps/translations/app_translations/README.md]: 👍

File: corehq/apps/translations/app_translations/upload_app.py
Comment (mkangia) [inline corehq/apps/translations/app_translations/upload_app.py]: nit: `could` not recognize?

File: corehq/apps/translations/app_translations/upload_form.py
Comment (mkangia) [inline corehq/apps/translations/app_translations/upload_form.py]: This is one method :(

**Review by mkangia — COMMENTED** [review]
Do you plan to break the longer methods in near future @orangejenny ?

This is definitely better with better collections of methods.. though we still have very long methods.

File: corehq/apps/translations/app_translations.py
Comment (orangejenny) [reply corehq/apps/translations/app_translations.py]: Not following this question - `get_unicode_dicts` is only called once on each sheet.

File: corehq/apps/translations/app_translations.py
Comment (orangejenny) [reply corehq/apps/translations/app_translations.py]: bc4b19e8672

File: corehq/apps/translations/app_translations.py
Comment (orangejenny) [reply corehq/apps/translations/app_translations.py]: bc4b19e8672

File: corehq/apps/translations/app_translations/upload_module.py
Comment (orangejenny) [reply corehq/apps/translations/app_translations/upload_module.py]: Leaving this one alone.

File: corehq/apps/translations/app_translations/README.md
Comment (orangejenny) [reply corehq/apps/translations/app_translations/README.md]: bcaf56b2d00

File: corehq/apps/translations/app_translations/upload_app.py
Comment (orangejenny) [reply corehq/apps/translations/app_translations/upload_app.py]: d6938292152

File: corehq/apps/translations/app_translations/upload_module.py
Comment (orangejenny) [reply corehq/apps/translations/app_translations/upload_module.py]: 902c98d4013

File: corehq/apps/translations/app_translations.py
Comment (orangejenny) [reply corehq/apps/translations/app_translations.py]: (resolved by later commit)

**Comment by kaapstorm** [comment]
Thanks @orangejenny

File: corehq/apps/translations/views.py
Comment (orangejenny) [reply corehq/apps/translations/views.py]: I don't know - this code is pre-existing.

File: corehq/apps/translations/app_translations.py
Comment (orangejenny) [reply corehq/apps/translations/app_translations.py]: 810cb075d2f

File: corehq/apps/translations/app_translations.py
Comment (orangejenny) [reply corehq/apps/translations/app_translations.py]: Not intentional, I must have done a bad rebase incorporating https://github.com/dimagi/commcare-hq/pull/23403

Fixing in 7150aa0e9fc

File: corehq/apps/translations/app_translations/upload_app.py
Comment (orangejenny) [reply corehq/apps/translations/app_translations/upload_app.py]: d32e3694a2e

File: corehq/apps/translations/app_translations.py
Comment (orangejenny) [reply corehq/apps/translations/app_translations.py]: I like the symmetry of `_check_for_workbook_error`, `_check_for_sheet_error`, and `_check_for_sheet_warnings`

File: corehq/apps/translations/app_translations.py
Comment (orangejenny) [reply corehq/apps/translations/app_translations.py]: also 72ca8681493

File: corehq/apps/translations/app_translations.py
Comment (mkangia) [reply corehq/apps/translations/app_translations.py]: you missed the second "r" in the fix.

**Comment by mkangia** [comment]
Build failures look related
```
======================================================================
ERROR: corehq.apps.translations.tests.test_bulk_app_translation:BulkAppTranslationBasicTest.test_bad_column_name
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/commcare-hq-ro/corehq/apps/translations/tests/test_bulk_app_translation.py", line 411, in test_bad_column_name
    'App Translations Updated!'
  File "/mnt/commcare-hq-ro/corehq/apps/translations/tests/test_bulk_app_translation.py", line 80, in upload_raw_excel_translations
    messages = process_bulk_app_translation_upload(self.app, workbook, expected_headers)
  File "/mnt/commcare-hq-ro/corehq/apps/translations/app_translations/upload_app.py", line 102, in process_bulk_app_translation_upload
    sheet_name=sheet.worksheet.title))
  File "/mnt/commcare-hq-ro/corehq/apps/translations/app_translations/upload_app.py", line 121, in _process_multi_sheet_rows
    return update_app_from_form_sheet(app, rows, identifier)
  File "/mnt/commcare-hq-ro/corehq/apps/translations/app_translations/upload_form.py", line 152, in update_app_from_form_sheet
    {'form': trans_type})
  File "/mnt/commcare-hq-ro/corehq/apps/translations/app_translations/upload_form.py", line 200, in _update_translation_node
    text_node.xml.append(e)
NameError: global name 'text_node' is not defined
----------------------------------------------------------------------
```

```
======================================================================
ERROR: corehq.apps.translations.tests.test_bulk_app_translation:AggregateMarkdownNodeTests.test_markdown_node
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/commcare-hq-ro/corehq/apps/translations/tests/test_bulk_app_translation.py", line 803, in test_markdown_node
    msgs = update_app_from_form_sheet(self.app, sheet, sheet.worksheet.title)
  File "/mnt/commcare-hq-ro/corehq/apps/translations/app_translations/upload_form.py", line 140, in update_app_from_form_sheet
    delete_node=(not keep_value_node)
  File "/mnt/commcare-hq-ro/corehq/apps/translations/app_translations/upload_form.py", line 200, in _update_translation_node
    text_node.xml.append(e)
NameError: global name 'text_node' is not defined
----------------------------------------------------------------------
```

File: corehq/apps/translations/app_translations.py
Comment (czue) [reply corehq/apps/translations/app_translations.py]: > `get_unicode_dicts` is only called once on each sheet

This answers the question - didn't realize that.

File: corehq/apps/translations/app_translations/upload_form.py
Comment (kaapstorm) [inline corehq/apps/translations/app_translations/upload_form.py]: I was curious what you were going to do with @mkangia's comment, @orangejenny, because I thought dropping the `r` prefixes was fine. :grin:

@mkangia you probably know this already, but just to confirm we're on the same page, the `r` prefix just means "[all backslashes are left in the string](https://docs.python.org/2.0/ref/strings.html)". The first pattern has no backslashes, so the `r` has no effect, and the second pattern has one, but `\s` is not an escape sequence, so the `r` has no effect there either.

... I was going to leave this comment at that, but I went looking for what Python 3 docs said about raw string literals, and kept coming across regex patterns that had `r` prefixes even though they clearly didn't need them because they didn't have any backslashes. [Then I found](https://docs.python.org/3/library/re.html#regular-expression-syntax) "... it’s highly recommended that you use raw strings for all but the simplest expressions". Hmmm. Maybe this is a convention that GitHub's syntax highlighting assumes and I was just unaware of. (In my defence, PyCharm uses regex syntax highlighting for all patterns, regardless of whether they are raw string literals or not.)

Anyway, fwiw, I'm fine with these patterns as they are. I'm also fine without the `r` prefix because they don't need them. But if using raw strings for all regex patterns is a thing I didn't know about, we should probably just always use them.

Comment (mkangia) [reply corehq/apps/translations/app_translations/upload_form.py]: Thanks for sharing that @kaapstorm. I just choose to not modify what i don't understand completely or whose intent is not clear to me. After reading your note i confused myself with somethings i tried with re.sub, so i am going to leave this to Jenny to decide to change or not.

Comment (orangejenny) [reply corehq/apps/translations/app_translations/upload_form.py]: Based on https://github.com/dimagi/commcare-hq/pull/23403 my understanding is that our convention is to use the `r` prefix for all regular expressions. @nickpell is that accurate?

File: corehq/apps/translations/app_translations/upload_form.py
Comment (orangejenny) [reply corehq/apps/translations/app_translations/upload_form.py]: Split up in a couple of commits: d3885475dc049f74bc1fa31177194ceab059b004 b589442e80342e01722bd18257549456fea28566 251a6320b43a65467bcc4ab855470b5427e9f14b 8e58c1567f6ae6cd1e1de0a6c658a0ef7222f40a

**Comment by orangejenny** [comment]
All code review feedback should now be addressed - thank you for all of the feedback.

@mkangia I'm not planning to do further refactoring, but let me know if there's anything that's still particularly egregious and I'll consider it ;)

@czue Good point about test coverage. It's relatively good in this area but could certainly be better, especially since I'm adding new code. I'll take a pass at adding tests.

File: corehq/apps/translations/app_translations/upload_form.py
Comment (mkangia) [reply corehq/apps/translations/app_translations/upload_form.py]: Thank you :)

**Review by mkangia — APPROVED** [review]
Thank you for stretching the scope of this PR for code improvement.

**Comment by czue** [comment]
thanks @orangejenny !

**Comment by orangejenny** [comment]
Added a few more refactoring commits, grouping some related helper functions into classes to reduce the amount of state they pass around.

Still working on increasing test coverage.

**Comment by orangejenny** [comment]
@czue Added test coverage.

**Comment by czue** [comment]
> @czue Added test coverage.

Hooray, thank you!

File: corehq/apps/translations/app_translations/README.md
Comment (emord) [inline corehq/apps/translations/app_translations/README.md]: should the modules in this line be called menus?

File: corehq/apps/translations/integrations/transifex/parser.py
Comment (emord) [inline corehq/apps/translations/integrations/transifex/parser.py]: should the other instances of module be changed to menu as well

**Review by emord — COMMENTED** [review]
Did a quick pass over this. I'm not quite sure where we're drawing the line between module and menus here

I think I may be missing it, but where are the xlsx files in app_manager being used? All of the tests I'm seeing use python objects to upload translations

File: corehq/apps/translations/app_translations/utils.py
Comment (orangejenny) [reply corehq/apps/translations/app_translations/utils.py]: Don't know what was up with that, don't think I added it. 9c494dd7b72

File: corehq/apps/app_manager/tests/data/bulk_app_translation/download/app.json
Comment (orangejenny) [reply corehq/apps/app_manager/tests/data/bulk_app_translation/download/app.json]: Unfortunately not - this change makes the download test pretty comprehensive, but many of the features being tested (e.g., graphs) aren't supported by AppFactory.

File: corehq/apps/translations/app_translations/README.md
Comment (orangejenny) [reply corehq/apps/translations/app_translations/README.md]: 240f7ac54ca

> From what I understand, only a subset of languages is supported without a feature flag

What do you mean? I think you can leave out rows. Maybe see if the update I made sounds accurate to you?

File: corehq/apps/translations/app_translations/README.md
Comment (orangejenny) [reply corehq/apps/translations/app_translations/README.md]: Eh, the code primarily uses "module" so I think it's fine if the docs do.

File: corehq/apps/translations/integrations/transifex/parser.py
Comment (orangejenny) [reply corehq/apps/translations/integrations/transifex/parser.py]: Are there particular ones you noticed? It's a bit hairy in this file because the transifex resources are, and should be, still named "module" - I don't want to be changing the names of anything in transifex. I'm fairly dependent on @mkangia to review these changes.

**Comment by orangejenny** [comment]
> I'm not quite sure where we're drawing the line between module and menus here

Everything user-facing in HQ should be menu; all of the code should be module. I don't know if the transifex resource names are user-facing in the transifex interface, but if they are, I don't care that they still say module - the "module" => "menu" change is more for SaaS users.

> I think I may be missing it, but where are the xlsx files in app_manager being used? All of the tests I'm seeing use python objects to upload translations

https://github.com/dimagi/commcare-hq/blob/ffd94d714fff2482f304afb85cafcf213e0f644c/corehq/apps/translations/tests/test_bulk_app_translation.py#L86

File: corehq/apps/translations/app_translations/README.md
Comment (emord) [reply corehq/apps/translations/app_translations/README.md]: I was thinking of this PR https://github.com/dimagi/commcare-hq/pull/23367

However maybe that doesn't apply here as those are "UI" translations and not in the app_translations folder?

**Comment by emord** [comment]
> Everything user-facing in HQ should be menu; all of the code should be module

Is there a technical reason for this other than changing it would be difficult? (not trying to sound judge-y, just want to understand where the separation is and the reason for it. I realize that change would be a slog and prone to errors)

File: corehq/apps/translations/generators.py
Comment (mkangia) [inline corehq/apps/translations/generators.py]: This line currently generates this `Module:module1:ad7059becbdcd3dfe155c116d5ac6837017b4e31` but after this change it would generate `Menu:menu1:ad7059becbdcd3dfe155c116d5ac6837017b4e31`. We would want to keep it `Module` and `module` only because a change in occurrence would cause the existing translations to get removed.

File: corehq/apps/translations/app_translations/README.md
Comment (orangejenny) [reply corehq/apps/translations/app_translations/README.md]: Yeah, that flag is only CommCare translations, not app translations.

File: corehq/apps/translations/generators.py
Comment (orangejenny) [reply corehq/apps/translations/generators.py]: thank you - https://github.com/dimagi/commcare-hq/pull/23924

**Comment by orangejenny** [comment]
Doing all further updates in separate PRs since this comment thread is so long. I'll continue to merge master into this branch; rebasing has gotten hairy.

File: corehq/apps/hqwebapp/forms.py
Comment (mkangia) [inline corehq/apps/hqwebapp/forms.py]: we would need to change this to also have the language field added in case `can_validate_app_translations` is true to support https://github.com/dimagi/commcare-hq/pull/23817 ?

File: corehq/apps/hqwebapp/forms.py
Comment (mkangia) [inline corehq/apps/hqwebapp/forms.py]: Looks like the choices is still all app langs but the initial value is "All" instead of the default language as set up for the validation feature. Would need to address that for the validation feature, I can take that in a follow up PR once this is merged.

File: corehq/apps/translations/views.py
Comment (mkangia) [inline corehq/apps/translations/views.py]: i think you removed request.file during master update. It should be added and later run in `validate_bulk_app_translation_upload` if you look at the current state of the method.

File: corehq/apps/translations/app_translations/utils.py
Comment (mkangia) [inline corehq/apps/translations/app_translations/utils.py]: trying to convince you for a better name here if possible.
menu_or_form would have values like `module1` and `module1_form2` right?
`menu_or_form` sounds like this would have values ['module', 'form']. I would push to rename this back to `sheet_name` only or `module_or_form_name` instead.

File: corehq/apps/translations/views.py
Comment (orangejenny) [reply corehq/apps/translations/views.py]: Are you sure? The way I read the code, `request.file` gets turned into `workbook` a few lines up, and then `workbook` is how the data is passed to `validate_bulk_app_translation_upload`.

File: corehq/apps/translations/app_translations/utils.py
Comment (orangejenny) [reply corehq/apps/translations/app_translations/utils.py]: I can change to `module_or_form_name`. I'd rather not change back to `sheet_name` because that doesn't make sense for the single-sheet format, where these strings aren't sheet names.

Comment (orangejenny) [reply corehq/apps/translations/app_translations/utils.py]: Rather, I'll change to `menu_or_form_name`.

Comment (orangejenny) [reply corehq/apps/translations/app_translations/utils.py]: ...although I just realized that `menu_or_form_name` sounds like that column should contain the name, not a string like `menu1`. It'd be more accurate to use `menu_or_form_slug`, but `slug` isn't a user-friendly name. So I agree that `menu_or_form` sounds like values would be either `menu` or `form` but I don't have a better name. Could do something really generic, like `item`, but that isn't really clear either.

File: corehq/apps/hqwebapp/forms.py
Comment (orangejenny) [reply corehq/apps/hqwebapp/forms.py]: The UI does now default to "All" and hides the validation checkbox. The validation shows up when you pick a language: http://g.recordit.co/3VbpJ6Z4nG.gif

I'm not totally in love with this UX, but we shouldn't have more than one language dropdown in the same form. If this UI is confusing, we should look at separating validation from uploading, rather than having validation be a checkbox.

Comment (mkangia) [reply corehq/apps/hqwebapp/forms.py]: Interesting work around there. Would just need to make people aware of this change, i am sure they would be searching for the validation option.

**Pull request merged**
